src/ripple/ledger/impl/ApplyStateTable.cpp

```diff
@@ -0,0 +1,589 @@
+        return nullptr;
+    return sle;
+}
+
+void
+ApplyStateTable::erase(
+    ReadView const& base,
+        std::shared_ptr<SLE> const& sle)
+{
+    auto const iter =
+        items_.find(sle->key());
+    if (iter == items_.end())
+        LogicError("ApplyStateTable::erase: missing key");
+    auto& item = iter->second;
+    if (item.second != sle)
+        LogicError("ApplyStateTable::erase: unknown SLE");
+    switch(item.first)
+    {
+    case Action::erase:
+        LogicError("ApplyStateTable::erase: double erase");
+        break;
+    case Action::insert:
+        items_.erase(iter);
+        break;
+    case Action::cache:
+    case Action::modify:
+        item.first = Action::erase;
+        break;
+    }
+}
+
+void
+ApplyStateTable::rawErase (ReadView const& base,
+    std::shared_ptr<SLE> const& sle)
+{
+    using namespace std;
+    auto const result = items_.emplace(
+        piecewise_construct,
+            forward_as_tuple(sle->key()),
+                forward_as_tuple(Action::erase, sle));
+    if (result.second)
+        return;
+    auto& item = result.first->second;
+    switch(item.first)
+    {
+    case Action::erase:
+        LogicError("ApplyStateTable::rawErase: double erase");
+        break;
+    case Action::insert:
+        items_.erase(result.first);
+        break;
+    case Action::cache:
+    case Action::modify:
+        item.first = Action::erase;
+        item.second = sle;
+        break;
+    }
+}
+
+void
+ApplyStateTable::insert (ReadView const& base,
+    std::shared_ptr<SLE> const& sle)
+{
+    auto const iter =
+        items_.lower_bound(sle->key());
+    if (iter == items_.end() ||
+        iter->first != sle->key())
+    {
+        using namespace std;
+        items_.emplace_hint(iter,
+            piecewise_construct,
+                forward_as_tuple(sle->key()),
+                    forward_as_tuple(Action::insert, sle));
+        return;
+    }
+    auto& item = iter->second;
+    switch(item.first)
+    {
+    case Action::cache:
+        LogicError("ApplyStateTable::insert: already cached");
+    case Action::insert:
+        LogicError("ApplyStateTable::insert: already inserted");        
+    case Action::modify:
+        LogicError("ApplyStateTable::insert: already modified");
+    case Action::erase:
+        break;
+    }
+    item.first = Action::modify;
+    item.second = sle;
+}
+
+void
+ApplyStateTable::replace (ReadView const& base,
+    std::shared_ptr<SLE> const& sle)
+{
+    auto const iter =
+        items_.lower_bound(sle->key());
+    if (iter == items_.end() ||
+        iter->first != sle->key())
+    {
+        using namespace std;
+        items_.emplace_hint(iter, piecewise_construct,
+            forward_as_tuple(sle->key()),
+                forward_as_tuple(Action::modify, sle));
+        return;
+    }
+    auto& item = iter->second;
+    switch (item.first)
+    {
+    case Action::erase:
+        LogicError("ApplyStateTable::replace: already erased");
```

Comment:
You can't replace something that doesn't exist. A replace is a logically different operation than creation and has very different rules in the ledger API.